### PR TITLE
[quant] Add quantize_per_channel and dequantize_per_channel operators to decomposition table

### PR DESCRIPTION
**Pull request by jerryzh168**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #94227

Summary:
same as https://github.com/pytorch/pytorch/pull/93312, but for quantize_per_channel and dequantize_per_channel

Test Plan:
python test/test_quantization.py TestQuantizePT2E.test_q_dq_per_channel_decomposition

Reviewers:

Subscribers:

Tasks:

Tags:

### CODE REVIEW

File: torch/ao/quantization/_pt2e/decompositions.py [hunk 1]
```diff
@@ -27,3 +27,56 @@ def dequantize_per_tensor(
...
+        axis: int,
+        quant_min: int,
+        quant_max: int,
+        dtype: torch.dtype
+) -> torch.Tensor:
+    input, permute_axis_list = _permute_to_axis_zero(input, axis)
+    res = torch.zeros_like(input)
+
+    for i in range(input.size(0)):
+        res[i] = torch.clamp(
```
Comment (vadimkantorov) [inline torch/ao/quantization/_pt2e/decompositions.py:53]: could this not be done without the loop if broadcast_right existed: https://github.com/pytorch/pytorch/issues/62148

Comment (jerryzh168) [reply torch/ao/quantization/_pt2e/decompositions.py:53]: yeah probably, broadcasting is typically implicit though, how do you call broadcast_right?

Comment (vadimkantorov) [reply torch/ao/quantization/_pt2e/decompositions.py:53]: it doesn't exist currently in python :( created a feature request here: https://github.com/pytorch/pytorch/issues/94233

what you could do is sth like `scales = scales.unflatten(-1, scales.shape[-1:] + (1, ) * (res.dim() - scales.dim()))` (though likely not torchscriptable if it matters here)

or `torch.round(input.movedim(0, -1) / scales) + zero_points, quant_min, quant_max).movedim(-1, 0)` - although not sure what it does to preserving memory format

File: torch/ao/quantization/_pt2e/decompositions.py [hunk 2]
```diff
@@ -27,3 +27,56 @@ def dequantize_per_tensor(
...
+def quantize_per_channel(
+        input: torch.Tensor,
+        scales: torch.Tensor,
+        zero_points: torch.Tensor,
+        axis: int,
+        quant_min: int,
+        quant_max: int,
+        dtype: torch.dtype
+) -> torch.Tensor:
+    input, permute_axis_list = _permute_to_axis_zero(input, axis)
```
Comment (vadimkantorov) [inline torch/ao/quantization/_pt2e/decompositions.py:49]: is it same as calling here input = input.transpose(axis, 0) and then call this again at the end of the function?

Comment (vadimkantorov) [reply torch/ao/quantization/_pt2e/decompositions.py:49]: So btw if this is instead dome as input.transpose(axis, -1) or even input.movedim(axis, -1) (and then the reverse at the end of the function), then regular broadcasting will work

Comment (jerryzh168) [reply torch/ao/quantization/_pt2e/decompositions.py:49]: yeah feels like this is the same, I can add this as a follow up PR to improve the implementation

File: torch/ao/quantization/_pt2e/decompositions.py (hunk 1 above)
Comment (jerryzh168) [reply torch/ao/quantization/_pt2e/decompositions.py:53]: this looks a bit complicated, I'll try transpose instead, thanks for the suggestions

Comment (vadimkantorov) [reply torch/ao/quantization/_pt2e/decompositions.py]: transpose or movedim is the cleanest, yes

**Comment by vadimkantorov** [comment]
@jerryzh168 Sorry, can't review fully, I'm not a maintainer, just hustling around in a way

File: torch/ao/quantization/fx/_decomposed.py [hunk 3]
```diff
@@ -27,6 +27,40 @@ def _dequantize_per_tensor_impl(
 ) -> torch.Tensor:
     return (input.to(torch.float32) - zero_point) * scale
 
+def _quantize_per_channel_impl(
+        input: torch.Tensor,
+        scales: torch.Tensor,
+        zero_points: torch.Tensor,
+        axis: int,
```
Comment (vadimkantorov) [inline torch/ao/quantization/fx/_decomposed.py:34]: is `dim` here better instead of `axis` to match better pytorch conventions?

Comment (jerryzh168) [reply torch/ao/quantization/fx/_decomposed.py:34]: yeah probably, maybe we can rename this later since this is used in existing quantize_per_channel ops right now as well

File: torch/ao/quantization/fx/_decomposed.py [hunk 4]
```diff
@@ -3,6 +3,40 @@
...
+        quant_min: int,
+        quant_max: int,
+        dtype: torch.dtype
+) -> torch.Tensor:
+    input = input.transpose(axis, -1)
+    output = torch.clamp(torch.round(input * (1.0 / scales)) + zero_points, quant_min, quant_max)
```
Comment (vkuzo) [inline torch/ao/quantization/fx/_decomposed.py:16]: nit: just divide by scales?

**Review by vkuzo — APPROVED** [review]

**Comment by vadimkantorov** [comment]
@vkuzo should this be merged or discarded?

**Comment by vkuzo** [comment]
> @vkuzo should this be merged or discarded?

@jerryzh168  would know!

**Comment by jerryzh168** [comment]
haven't decided but I think it's likely going to be merged

**Comment by jerryzh168** [comment]
doing this in a different PR: https://github.com/pytorch/pytorch/pull/105783
